File: sdks/ts/src/queues.ts
import type { Client } from "@connectrpc/connect";
import { create } from "@bufbuild/protobuf";
import {
  type ErrorInfo,
  Gateway,
  Header as ProtoHeader,
  QueueAckEntrySchema,
  type QueueAckResult as ProtoQueueAckResult,
  QueueAckType,
  QueueAcquireRequestSchema,
  QueueAcknowledgeRequestSchema,
  QueueRenewRequestSchema,
} from "../gen/crabka/gateway/v1/gateway_pb.ts";
import { fromConnectError, fromRecordError, type ErrorKind, InvalidArgumentError, TransportError } from "./errors.ts";
import type { Header, MockStore, StoredMessage, StoredQueueDelivery } from "./messaging.ts";

const DEFAULT_QUEUE_LOCK_DURATION_MS = 30_000;
const QUEUE_MESSAGE_ID_PARTS = 3;
const QUEUE_MESSAGE_NOT_ACQUIRED = "queue message is not acquired";
const QUEUE_SESSION_EXPIRED = "queue session expired; re-acquire";

type LiveGateway = Client<typeof Gateway>;

export type QueueAckTypeName = "accept" | "release" | "reject";
export type QueueAckEntry = { messageId: string; ackType: QueueAckTypeName };
export type QueueRenewEntry = { messageId: string };
export type QueueOperationError = { kind: ErrorKind; message: string; retriable: boolean };
export type QueueResult = { messageId: string; error: QueueOperationError | null };
export type QueueMessage = {
  messageId: string;
  topic: string;
  partition: number;
  offset: bigint;
  value?: Uint8Array;
  headers: Header[];
  deliveryCount: number;
};
export type QueueAcquireResult = { sessionId: string; messages: QueueMessage[] };
export type QueueBatchResult = { results: QueueResult[] };

export type QueueAcquireOptions = { group?: string; max?: number; lockDurationMs?: number; sessionId?: string };
export type QueuesModule = {
  acquire(topic: string, options: QueueAcquireOptions): Promise<QueueAcquireResult>;
  acknowledge(sessionId: string, entries: QueueAckEntry[]): Promise<QueueBatchResult>;
  renew(sessionId: string, entries: QueueRenewEntry[]): Promise<QueueBatchResult>;
};

export function createQueuesModule(options: {
  endpoint: string;
  gateway?: LiveGateway;
  headers?: HeadersInit;
  mockStore?: MockStore;
}): QueuesModule {
  return {
    acquire: (topic, acquireOptions) => acquire(options, topic, acquireOptions),
    acknowledge: (sessionId, entries) => acknowledge(options, sessionId, entries),
    renew: (sessionId, entries) => renew(options, sessionId, entries),
  };
}

async function acquire(
  options: Parameters<typeof createQueuesModule>[0],
  topic: string,
  acquireOptions: QueueAcquireOptions,
): Promise<QueueAcquireResult> {
  if (!topic) {
    throw new InvalidArgumentError("queue topic is required");
  }
  assertSupportedAcquireOptions(acquireOptions);
  if (options.endpoint.startsWith("unreachable://")) {
    throw new TransportError("endpoint unreachable");
  }
  if (options.mockStore) {
    return acquireMockMessages(
      options.mockStore,
      topic,
      acquireOptions.group ?? "",
      acquireOptions.max ?? 1,
      acquireOptions.sessionId,
    );
  }
  if (!options.gateway) {
    throw new TransportError("gateway transport is not configured");
  }

  try {
    const response = await options.gateway.queueAcquire(
      create(QueueAcquireRequestSchema, {
        groupId: acquireOptions.group,
        topics: [topic],
        maxMessages: acquireOptions.max ?? 1,
        waitMs: 0,
        sessionId: acquireOptions.sessionId,
        lockDurationMs: BigInt(DEFAULT_QUEUE_LOCK_DURATION_MS),
      }),
      { headers: options.headers },
    );
    return {
      sessionId: response.sessionId,
      messages: response.messages.map((message) => ({
        messageId: messageIdFromParts(message.topic, message.partition, message.offset),
        topic: message.topic,
        partition: message.partition,
        offset: message.offset,
        value: message.value,
        headers: fromProtoHeaders(message.headers),
        deliveryCount: message.deliveryCount,
      })),
    };
  } catch (error) {
    throw fromConnectError(error);
  }
}

async function acknowledge(
  options: Parameters<typeof createQueuesModule>[0],
  sessionId: string,
  entries: QueueAckEntry[],
): Promise<QueueBatchResult> {
  assertSessionId(sessionId);
  entries.forEach((entry) => assertAckType(entry.ackType));
  if (options.mockStore) {
    const store = options.mockStore;
    const group = assertMockSession(store, sessionId).group;
    return { results: entries.map((entry) => acknowledgeMockMessage(store, group, sessionId, entry)) };
  }
  if (!options.gateway) {
    throw new TransportError("gateway transport is not configured");
  }

  try {
    const response = await options.gateway.queueAcknowledge(
      create(QueueAcknowledgeRequestSchema, { sessionId, entries: entries.map(toProtoAckEntry) }),
      { headers: options.headers },
    );
    return { results: response.results.map(fromProtoQueueResult) };
  } catch (error) {
    throw fromConnectError(error);
  }
}

async function renew(
  options: Parameters<typeof createQueuesModule>[0],
  sessionId: string,
  entries: QueueRenewEntry[],
): Promise<QueueBatchResult> {
  assertSessionId(sessionId);
  if (options.mockStore) {
    const store = options.mockStore;
    const group = assertMockSession(store, sessionId).group;
    return { results: entries.map((entry) => renewMockMessage(store, group, sessionId, entry)) };
  }
  if (!options.gateway) {
    throw new TransportError("gateway transport is not configured");
  }

  try {
    const response = await options.gateway.queueRenew(
      create(QueueRenewRequestSchema, {
        sessionId,
        entries: entries.map((entry) => toProtoAckEntry({ messageId: entry.messageId, ackType: "accept" })),
      }),
      { headers: options.headers },
    );
    return { results: response.results.map(fromProtoQueueResult) };
  } catch (error) {
    throw fromConnectError(error);
  }
}

function assertSupportedAcquireOptions(options: QueueAcquireOptions): void {
  if (!options.group) {
    throw new InvalidArgumentError("queue group is required");
  }
  if (options.lockDurationMs !== undefined && options.lockDurationMs !== DEFAULT_QUEUE_LOCK_DURATION_MS) {
    throw new InvalidArgumentError(
      "queue lock_duration_ms must be 30000; per-acquire lock durations are not supported",
    );
  }
}

function assertSessionId(sessionId: string): void {
  if (sessionId === "") {
    throw new InvalidArgumentError("queue session_id is required");
  }
}

function assertAckType(ackType: string): void {
  if (ackType !== "accept" && ackType !== "release" && ackType !== "reject") {
    throw new InvalidArgumentError("queue ack_type must be accept, release, or reject");
  }
}

function acquireMockMessages(
  store: MockStore,
  topic: string,
  group: string,
  maxMessages: number,
  sessionId?: string,
): QueueAcquireResult {
  const effectiveMax = Math.min(Math.max(maxMessages, 1), 500);
  if (!sessionId) {
    sessionId = `queue-session-${store.nextQueueSessionId}`;
    store.nextQueueSessionId += 1;
    store.queueSessions.set(sessionId, { topic, group, max: effectiveMax });
  } else {
    const session = assertMockSession(store, sessionId);
    if (session.topic !== topic || session.group !== group) {
      throw new InvalidArgumentError("group_id and topics are fixed when a queue session is created");
    }
    if (maxMessages !== 0 && effectiveMax !== session.max) {
      throw new InvalidArgumentError("max_messages is fixed when a queue session is created");
    }
  }
  const messages = store.messages
    .filter((message) => message.record.topic === topic && mockDelivery(message, group).state === "available")
    .slice(0, effectiveMax)
    .map((message) => acquireMockMessage(message, group, sessionId));
  return { sessionId, messages };
}

function acquireMockMessage(message: StoredMessage, group: string, sessionId: string): QueueMessage {
  const delivery = mockDelivery(message, group);
  delivery.state = "acquired";
  delivery.sessionId = sessionId;
  delivery.deliveryCount += 1;
  return {
    messageId: messageIdFromStoredMessage(message),
    topic: message.record.topic,
    partition: message.partition,
    offset: message.offset,
    value: new Uint8Array(message.record.value),
    headers: message.record.headers.map((header) => ({
      name: header.name,
      value: header.value ? new Uint8Array(header.value) : undefined,
    })),
    deliveryCount: delivery.deliveryCount,
  };
}

function acknowledgeMockMessage(store: MockStore, group: string, sessionId: string, entry: QueueAckEntry): QueueResult {
  const delivery = findAcquiredMockMessage(store, group, sessionId, entry.messageId);
  if (!delivery) {
    return queueEntryError(entry.messageId);
  }
  delivery.state = mockQueueStateForAck(entry.ackType);
  delete delivery.sessionId;
  return { messageId: entry.messageId, error: null };
}

function renewMockMessage(store: MockStore, group: string, sessionId: string, entry: QueueRenewEntry): QueueResult {
  if (!findAcquiredMockMessage(store, group, sessionId, entry.messageId)) {
    return queueEntryError(entry.messageId);
  }
  return { messageId: entry.messageId, error: null };
}

function findAcquiredMockMessage(
  store: MockStore,
  group: string,
  sessionId: string,
  messageId: string,
): StoredQueueDelivery | undefined {
  const message = store.messages.find((candidate) => messageIdFromStoredMessage(candidate) === messageId);
  const delivery = message?.queueDeliveries.get(group);
  return delivery?.state === "acquired" && delivery.sessionId === sessionId ? delivery : undefined;
}

function mockDelivery(message: StoredMessage, group: string): StoredQueueDelivery {
  let delivery = message.queueDeliveries.get(group);
  if (!delivery) {
    delivery = { state: "available", deliveryCount: 0 };
    message.queueDeliveries.set(group, delivery);
  }
  return delivery;
}

function assertMockSession(store: MockStore, sessionId: string): { topic: string; group: string; max: number } {
  const session = store.queueSessions.get(sessionId);
  if (!session) {
    throw new InvalidArgumentError(QUEUE_SESSION_EXPIRED);
  }
  return session;
}

function mockQueueStateForAck(ackType: QueueAckTypeName): StoredQueueDelivery["state"] {
  if (ackType === "release") {
    return "available";
  }
  if (ackType === "reject") {
    return "rejected";
  }
  return "accepted";
}

function queueEntryError(messageId: string): QueueResult {
  return { messageId, error: { kind: "invalid_argument", message: QUEUE_MESSAGE_NOT_ACQUIRED, retriable: false } };
}

function toProtoAckEntry(entry: QueueAckEntry) {
  const messageId = parseMessageId(entry.messageId);
  return create(QueueAckEntrySchema, {
    topic: messageId.topic,
    partition: messageId.partition,
    offset: messageId.offset,
    type: toProtoAckType(entry.ackType),
  });
}

function parseMessageId(messageId: string): { topic: string; partition: number; offset: bigint } {
  const parts = messageId.split(":");
  if (parts.length !== QUEUE_MESSAGE_ID_PARTS) {
    throw new InvalidArgumentError("queue message_id must be <topic>:<partition>:<offset>");
  }
  const [topic, partitionText, offsetText] = parts;
  if (topic === undefined || partitionText === undefined || offsetText === undefined) {
    throw new InvalidArgumentError("queue message_id must be <topic>:<partition>:<offset>");
  }
  if (!/^-?\d+$/.test(partitionText) || !/^-?\d+$/.test(offsetText)) {
    throw new InvalidArgumentError("queue message_id must be <topic>:<partition>:<offset>");
  }
  const partition = Number(partitionText);
  let offset: bigint;
  try {
    offset = BigInt(offsetText);
  } catch {
    throw new InvalidArgumentError("queue message_id must be <topic>:<partition>:<offset>");
  }
  if (!topic || !Number.isInteger(partition) || partition < -2_147_483_648 || partition > 2_147_483_647) {
    throw new InvalidArgumentError("queue message_id must be <topic>:<partition>:<offset>");
  }
  return { topic, partition, offset };
}

function toProtoAckType(ackType: QueueAckTypeName): QueueAckType {
  if (ackType === "release") {
    return QueueAckType.RELEASE;
  }
  if (ackType === "reject") {
    return QueueAckType.REJECT;
  }
  return QueueAckType.ACCEPT;
}

function fromProtoQueueResult(result: ProtoQueueAckResult): QueueResult {
  if (!result.entry) {
    throw new TransportError("queue response result did not include an entry");
  }
  const messageId = messageIdFromParts(result.entry.topic, result.entry.partition, result.entry.offset);
  if (!result.error) {
    return { messageId, error: null };
  }
  return { messageId, error: queueOperationErrorFromProto(result.error) };
}

function queueOperationErrorFromProto(error: ErrorInfo): QueueOperationError {
  const mappedError = fromRecordError(error.code, error.message, error.retriable);
  return { kind: mappedError.kind, message: mappedError.message, retriable: error.retriable };
}

function fromProtoHeaders(headers: ProtoHeader[]): Header[] {
  return headers.map((header) => ({ name: header.key, value: header.value }));
}

function messageIdFromStoredMessage(message: StoredMessage): string {
  return messageIdFromParts(message.record.topic, message.partition, message.offset);
}

function messageIdFromParts(topic: string, partition: number, offset: bigint): string {
  return `${topic}:${partition}:${offset}`;
}

File: sdks/go/queues.go
package crabka

import (
	"context"
	"fmt"
	"strconv"
	"strings"

	"connectrpc.com/connect"
	gw "github.com/robot-head/crabka/sdks/go/gen/crabka/gateway/v1"
)

const (
	defaultQueueLockDurationMS int64 = 30_000
	queueMessageIDParts              = 3
	queueMessageNotAcquired          = "queue message is not acquired"
	queueSessionExpired              = "queue session expired; re-acquire"
)

type QueueAckType string

const (
	QueueAckAccept  QueueAckType = "accept"
	QueueAckRelease QueueAckType = "release"
	QueueAckReject  QueueAckType = "reject"
)

type QueueAckEntry struct {
	MessageID string
	AckType   QueueAckType
}

type QueueRenewEntry struct {
	MessageID string
}

type QueueOperationError struct {
	Kind      ErrorKind
	Message   string
	Retriable bool
}

type QueueResult struct {
	MessageID string
	Error     *QueueOperationError
}

type QueueMessage struct {
	MessageID     string
	Topic         string
	Partition     int32
	Offset        int64
	Value         []byte
	Headers       []Header
	DeliveryCount int32
}

type QueueAcquireResult struct {
	SessionID string
	Messages  []QueueMessage
}

type QueueBatchResult struct {
	Results []QueueResult
}

type Queues struct{ client *Client }

type parsedQueueMessageID struct {
	topic     string
	partition int32
	offset    int64
}

func (q *Queues) Acquire(ctx context.Context, topic string, group string, max uint32, lockDurationMS int64) (QueueAcquireResult, error) {
	return q.AcquireWithSession(ctx, topic, group, max, lockDurationMS, "")
}

func (q *Queues) AcquireWithSession(ctx context.Context, topic string, group string, max uint32, lockDurationMS int64, sessionID string) (QueueAcquireResult, error) {
	if err := validateQueueAcquire(topic, group, lockDurationMS); err != nil {
		return QueueAcquireResult{}, err
	}
	if strings.HasPrefix(q.client.endpoint, "unreachable://") {
		return QueueAcquireResult{}, errorWithMessage(Transport, "endpoint unreachable")
	}
	if q.client.mockStore != nil {
		return q.client.mockStore.acquireQueueMessages(topic, group, max, sessionID)
	}
	request := connect.NewRequest(&gw.QueueAcquireRequest{GroupId: group, Topics: []string{topic}, MaxMessages: max, WaitMs: 0, SessionId: sessionID, LockDurationMs: uint64(defaultQueueLockDurationMS)})
	addAuthorization(request.Header(), q.client.bearerToken)
	response, err := q.client.gateway.QueueAcquire(ctx, request)
	if err != nil {
		return QueueAcquireResult{}, mapConnectError(err)
	}
	return fromProtoQueueAcquire(response.Msg), nil
}

func (q *Queues) Acknowledge(ctx context.Context, sessionID string, entries []QueueAckEntry) (QueueBatchResult, error) {
	if err := validateQueueSessionID(sessionID); err != nil {
		return QueueBatchResult{}, err
	}
	if err := validateQueueAckTypes(entries); err != nil {
		return QueueBatchResult{}, err
	}
	if q.client.mockStore != nil {
		return q.client.mockStore.acknowledgeQueueMessages(sessionID, entries)
	}
	protoEntries, err := toProtoQueueAckEntries(entries)
	if err != nil {
		return QueueBatchResult{}, err
	}
	request := connect.NewRequest(&gw.QueueAcknowledgeRequest{SessionId: sessionID, Entries: protoEntries})
	addAuthorization(request.Header(), q.client.bearerToken)
	response, err := q.client.gateway.QueueAcknowledge(ctx, request)
	if err != nil {
		return QueueBatchResult{}, mapConnectError(err)
	}
	return fromProtoQueueBatch(response.Msg.Results)
}

func (q *Queues) Renew(ctx context.Context, sessionID string, entries []QueueRenewEntry) (QueueBatchResult, error) {
	if err := validateQueueSessionID(sessionID); err != nil {
		return QueueBatchResult{}, err
	}
	if q.client.mockStore != nil {
		return q.client.mockStore.renewQueueMessages(sessionID, entries)
	}
	protoEntries, err := toProtoQueueRenewEntries(entries)
	if err != nil {
		return QueueBatchResult{}, err
	}
	request := connect.NewRequest(&gw.QueueRenewRequest{SessionId: sessionID, Entries: protoEntries})
	addAuthorization(request.Header(), q.client.bearerToken)
	response, err := q.client.gateway.QueueRenew(ctx, request)
	if err != nil {
		return QueueBatchResult{}, mapConnectError(err)
	}
	return fromProtoQueueBatch(response.Msg.Results)
}

func (q *Queues) Ack(ctx context.Context, messageID string) error {
	result, err := q.Acknowledge(ctx, "legacy-ack", []QueueAckEntry{{MessageID: messageID, AckType: QueueAckAccept}})
	if err != nil {
		return err
	}
	if len(result.Results) == 0 || result.Results[0].Error == nil {
		return nil
	}
	return errorWithMessage(result.Results[0].Error.Kind, result.Results[0].Error.Message)
}

func validateQueueAcquire(topic string, group string, lockDurationMS int64) error {
	if topic == "" {
		return errorWithMessage(InvalidArgument, "queue topic is required")
	}
	if group == "" {
		return errorWithMessage(InvalidArgument, "queue group is required")
	}
	if lockDurationMS != 0 && lockDurationMS != defaultQueueLockDurationMS {
		return errorWithMessage(InvalidArgument, "queue lock_duration_ms must be 30000; per-acquire lock durations are not supported")
	}
	return nil
}

func validateQueueSessionID(sessionID string) error {
	if sessionID == "" {
		return errorWithMessage(InvalidArgument, "queue session_id is required")
	}
	return nil
}

func validateQueueAckTypes(entries []QueueAckEntry) error {
	for _, entry := range entries {
		switch entry.AckType {
		case QueueAckAccept, QueueAckRelease, QueueAckReject:
		default:
			return errorWithMessage(InvalidArgument, "queue ack_type must be accept, release, or reject")
		}
	}
	return nil
}

func (s *mockStore) acquireQueueMessages(topic string, group string, max uint32, sessionID string) (QueueAcquireResult, error) {
	effectiveMax := max
	if effectiveMax == 0 {
		effectiveMax = 1
	}
	if effectiveMax > 500 {
		effectiveMax = 500
	}
	if sessionID == "" {
		sessionID = fmt.Sprintf("queue-session-%d", s.nextQueueSessionID)
		s.nextQueueSessionID++
		s.queueSessions[sessionID] = mockQueueSession{topic: topic, group: group, max: effectiveMax}
	} else {
		session, ok := s.queueSessions[sessionID]
		if !ok {
			return QueueAcquireResult{}, errorWithMessage(InvalidArgument, queueSessionExpired)
		}
		if session.topic != topic || session.group != group {
			return QueueAcquireResult{}, errorWithMessage(InvalidArgument, "group_id and topics are fixed when a queue session is created")
		}
		if max != 0 && effectiveMax != session.max {
			return QueueAcquireResult{}, errorWithMessage(InvalidArgument, "max_messages is fixed when a queue session is created")
		}
	}
	messages := make([]QueueMessage, 0, effectiveMax)
	for index := range s.messages {
		if uint32(len(messages)) == effectiveMax {
			break
		}
		message := &s.messages[index]
		if message.record.Topic != topic || message.queueDelivery(group).state != mockQueueMessageAvailable {
			continue
		}
		messages = append(messages, acquireMockQueueMessage(message, group, sessionID))
	}
	return QueueAcquireResult{SessionID: sessionID, Messages: messages}, nil
}

func acquireMockQueueMessage(message *mockMessage, group string, sessionID string) QueueMessage {
	delivery := message.queueDelivery(group)
	delivery.state = mockQueueMessageAcquired
	delivery.sessionID = sessionID
	delivery.deliveryCount++
	return QueueMessage{MessageID: queueMessageID(message.record.Topic, message.partition, message.offset), Topic: message.record.Topic, Partition: message.partition, Offset: message.offset, Value: append([]byte(nil), message.record.Value...), Headers: cloneHeaders(message.record.Headers), DeliveryCount: delivery.deliveryCount}
}

func (s *mockStore) acknowledgeQueueMessages(sessionID string, entries []QueueAckEntry) (QueueBatchResult, error) {
	if _, ok := s.queueSessions[sessionID]; !ok {
		return QueueBatchResult{}, errorWithMessage(InvalidArgument, queueSessionExpired)
	}
	results := make([]QueueResult, 0, len(entries))
	for _, entry := range entries {
		results = append(results, s.acknowledgeQueueMessage(sessionID, entry))
	}
	return QueueBatchResult{Results: results}, nil
}

func (s *mockStore) acknowledgeQueueMessage(sessionID string, entry QueueAckEntry) QueueResult {
	delivery := s.findAcquiredQueueMessage(sessionID, entry.MessageID)
	if delivery == nil {
		return queueEntryError(entry.MessageID)
	}
	delivery.state = mockQueueStateForAck(entry.AckType)
	delivery.sessionID = ""
	return QueueResult{MessageID: entry.MessageID}
}

func (s *mockStore) renewQueueMessages(sessionID string, entries []QueueRenewEntry) (QueueBatchResult, error) {
	if _, ok := s.queueSessions[sessionID]; !ok {
		return QueueBatchResult{}, errorWithMessage(InvalidArgument, queueSessionExpired)
	}
	results := make([]QueueResult, 0, len(entries))
	for _, entry := range entries {
		if s.findAcquiredQueueMessage(sessionID, entry.MessageID) == nil {
			results = append(results, queueEntryError(entry.MessageID))
			continue
		}
		results = append(results, QueueResult{MessageID: entry.MessageID})
	}
	return QueueBatchResult{Results: results}, nil
}

func (s *mockStore) findAcquiredQueueMessage(sessionID string, messageID string) *mockQueueDelivery {
	group := s.queueSessions[sessionID].group
	for index := range s.messages {
		message := &s.messages[index]
		delivery := message.queueDeliveries[group]
		if delivery != nil && delivery.state == mockQueueMessageAcquired && delivery.sessionID == sessionID && queueMessageID(message.record.Topic, message.partition, message.offset) == messageID {
			return delivery
		}
	}
	return nil
}

func mockQueueStateForAck(ackType QueueAckType) mockQueueMessageState {
	if ackType == QueueAckRelease {
		return mockQueueMessageAvailable
	}
	if ackType == QueueAckReject {
		return mockQueueMessageRejected
	}
	return mockQueueMessageAccepted
}

func queueEntryError(messageID string) QueueResult {
	return QueueResult{MessageID: messageID, Error: &QueueOperationError{Kind: InvalidArgument, Message: queueMessageNotAcquired}}
}

func toProtoQueueAckEntries(entries []QueueAckEntry) ([]*gw.QueueAckEntry, error) {
	protoEntries := make([]*gw.QueueAckEntry, 0, len(entries))
	for _, entry := range entries {
		protoEntry, err := toProtoQueueAckEntry(entry)
		if err != nil {
			return nil, err
		}
		protoEntries = append(protoEntries, protoEntry)
	}
	return protoEntries, nil
}

func toProtoQueueRenewEntries(entries []QueueRenewEntry) ([]*gw.QueueAckEntry, error) {
	protoEntries := make([]*gw.QueueAckEntry, 0, len(entries))
	for _, entry := range entries {
		protoEntry, err := toProtoQueueAckEntry(QueueAckEntry{MessageID: entry.MessageID, AckType: QueueAckAccept})
		if err != nil {
			return nil, err
		}
		protoEntries = append(protoEntries, protoEntry)
	}
	return protoEntries, nil
}

func toProtoQueueAckEntry(entry QueueAckEntry) (*gw.QueueAckEntry, error) {
	messageID, err := parseQueueMessageID(entry.MessageID)
	if err != nil {
		return nil, err
	}
	return &gw.QueueAckEntry{Topic: messageID.topic, Partition: messageID.partition, Offset: messageID.offset, Type: toProtoQueueAckType(entry.AckType)}, nil
}

func toProtoQueueAckType(ackType QueueAckType) gw.QueueAckType {
	if ackType == QueueAckRelease {
		return gw.QueueAckType_RELEASE
	}
	if ackType == QueueAckReject {
		return gw.QueueAckType_REJECT
	}
	return gw.QueueAckType_ACCEPT
}

func parseQueueMessageID(messageID string) (parsedQueueMessageID, error) {
	parts := strings.Split(messageID, ":")
	if len(parts) != queueMessageIDParts || parts[0] == "" {
		return parsedQueueMessageID{}, invalidQueueMessageIDError()
	}
	partition, err := strconv.ParseInt(parts[1], 10, 32)
	if err != nil {
		return parsedQueueMessageID{}, invalidQueueMessageIDError()
	}
	offset, err := strconv.ParseInt(parts[2], 10, 64)
	if err != nil {
		return parsedQueueMessageID{}, invalidQueueMessageIDError()
	}
	return parsedQueueMessageID{topic: parts[0], partition: int32(partition), offset: offset}, nil
}

func invalidQueueMessageIDError() error {
	return errorWithMessage(InvalidArgument, "queue message_id must be <topic>:<partition>:<offset>")
}

func fromProtoQueueAcquire(response *gw.QueueAcquireResponse) QueueAcquireResult {
	messages := make([]QueueMessage, 0, len(response.Messages))
	for _, message := range response.Messages {
		messages = append(messages, fromProtoQueuedMessage(message))
	}
	return QueueAcquireResult{SessionID: response.SessionId, Messages: messages}
}

func fromProtoQueuedMessage(message *gw.QueuedMessage) QueueMessage {
	return QueueMessage{MessageID: queueMessageID(message.Topic, message.Partition, message.Offset), Topic: message.Topic, Partition: message.Partition, Offset: message.Offset, Value: append([]byte(nil), message.Value...), Headers: fromProtoHeaders(message.Headers), DeliveryCount: message.DeliveryCount}
}

func fromProtoHeaders(headers []*gw.Header) []Header {
	converted := make([]Header, 0, len(headers))
	for _, header := range headers {
		converted = append(converted, Header{Name: header.Key, Value: cloneNullableBytes(header.Value)})
	}
	return converted
}

func fromProtoQueueBatch(results []*gw.QueueAckResult) (QueueBatchResult, error) {
	converted := make([]QueueResult, 0, len(results))
	for _, result := range results {
		if result == nil || result.Entry == nil {
			return QueueBatchResult{}, errorWithMessage(Transport, "queue response result did not include an entry")
		}
		converted = append(converted, fromProtoQueueResult(result))
	}
	return QueueBatchResult{Results: converted}, nil
}

func fromProtoQueueResult(result *gw.QueueAckResult) QueueResult {
	messageID := queueMessageID(result.Entry.Topic, result.Entry.Partition, result.Entry.Offset)
	if result.Error == nil {
		return QueueResult{MessageID: messageID}
	}
	return QueueResult{MessageID: messageID, Error: fromProtoQueueOperationError(result.Error)}
}

func fromProtoQueueOperationError(errorInfo *gw.ErrorInfo) *QueueOperationError {
	if errorInfo == nil {
		return nil
	}
	return &QueueOperationError{Kind: mapGatewayErrorKind(errorInfo.Code, errorInfo.Retriable), Message: errorInfo.Message, Retriable: errorInfo.Retriable}
}

func queueMessageID(topic string, partition int32, offset int64) string {
	return fmt.Sprintf("%s:%d:%d", topic, partition, offset)
}
